### PR TITLE
Add minimum dependency testing to CI

### PR DESCRIPTION
**Pull request by charlesbluca**
Continuing @jrbourbeau's work in #4794, this PR adds a new check that runs the test suite against the minimum versions of Distributed's dependencies, which should help us to identify if/when deps need to be updated.

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: .github/workflows/additional.yml
Comment (jrbourbeau) [inline .github/workflows/additional.yml:1]: This file looks very similar to the existing `tests.yaml` file. Perhaps it would be straightforward to just add a new build matrix entry for `mindeps` over in `tests.yaml`. What do you think?

File: continuous_integration/environment-mindeps.yaml
Comment (jrbourbeau) [inline continuous_integration/environment-mindeps.yaml:10]: Now's probably a good time to choose a minimum version for `jinja2`, `pyyaml`, and `urllib3`

**Review by jrbourbeau — COMMENTED** [review]
Thanks for your work here @charlesbluca!

File: continuous_integration/environment-mindeps.yaml
Comment (charlesbluca) [reply continuous_integration/environment-mindeps.yaml:10]: Yup, was going to start off with whatever version first added support for 3.8 (assuming the packages aren't `noarch`) and work from there

File: .github/workflows/additional.yml
Comment (charlesbluca) [reply .github/workflows/additional.yml:1]: That sounds reasonable to me! So far think the major differences here would be:

- skipping over `host_info.py` since the benchmark there requires numpy
- maybe handling caching a little differently since we theoretically shouldn't need to update the mindeps cache very often

Comment (jrbourbeau) [reply .github/workflows/additional.yml:1]: > skipping over host_info.py since the benchmark there requires numpy

Ah, good point. Adding a (psuedocode) `if: not mindeps` to that step seems nice and lightweight 

> maybe handling caching a little differently since we theoretically shouldn't need to update the mindeps cache very often

Sorry, I don't quite follow here. Why would environment caching need to differ for the mindeps build vs. "normal" test runs?

File: .github/workflows/tests.yaml
Comment (jrbourbeau) [inline .github/workflows/tests.yaml:156]:
```suggestion
        if: matrix.environment != 'mindeps'
```

Comment (charlesbluca) [reply .github/workflows/tests.yaml]: Thanks! Just caught this

File: .github/workflows/additional.yml
Comment (charlesbluca) [reply .github/workflows/additional.yml]: > Sorry, I don't quite follow here. Why would environment caching need to differ for the mindeps build vs. "normal" test runs?

Right now, we're constructing the conda environment cache key using the current date, so that regardless of changes to the conda environment files, the environment is reconstructed at least once a day:

https://github.com/dask/distributed/blob/27a91dd545d3638b6c3a6681089f0ceb91ed6b7d/.github/workflows/tests.yaml#L89-L104

For the test runs with unpinned dependencies, this is important because we want to regularly pull in the latest versions, but for the mindeps runs this isn't as much of an issue since the majority of the packages we're pulling in are years old are unlikely to be receiving new builds (outside of the `pytest-*` packages, which are unpinned). Thus, we could potentially change the logic for these runs so that the cached environment is kept for longer, which would speed things up a bit.

I haven't made the change yet as I wanted to make sure we aren't risking unexpected failures by caching the `pytest-*` packages, but if you think this is a reasonable change happy to do that as well 🙂

**Comment by jrbourbeau** [comment]
Looks like the `ci1` half of the `mindeps` build is passing 🎉

File: requirements.txt
Comment (jrbourbeau) [inline requirements.txt:4]: I don't know that we want to add `cytoolz` as a hard dependency here. I'm curious what the motivation behind this change is

Comment (charlesbluca) [reply requirements.txt:4]: Added it because it was listed as a runtime dependency in the recipe:

https://github.com/dask/distributed/blob/5b6351a6800b0cad76d599962d3394677c791d1a/continuous_integration/recipes/distributed/meta.yaml#L34

Though fi there's a reason that this should be included in the conda recipe but not here I can remove

Comment (eriknw) [reply requirements.txt:3]: Yeah, we now build _97_ wheels for `cytoolz`, so it should be reliable to install via pip.

Comment (charlesbluca) [reply requirements.txt:3]: > I think it was historically difficult to install cytoolz as it required a compiler to be installed

Ah okay, I had just checked to see if it was available on PyPI but didn't realized that older versions only publish the source distribution.

> we now build 97 wheels for cytoolz

It looks like wheels for cpython 3.8-3.10 are only available for 0.12.0 and up, so maybe we should hold off on adding here until distributed requires `cytoolz >=0.12.0`?

Comment (jrbourbeau) [reply requirements.txt:3]: > we now build 97 wheels

!

> It looks like wheels for cpython 3.8-3.10 are only available for 0.12.0 and up, so maybe we should hold off on adding here until distributed requires cytoolz >=0.12.0?

+1 for holding off for now. Let's keep as an optional dependency, but not a hard requirement

File: distributed/system_monitor.py
Comment (charlesbluca) [inline distributed/system_monitor.py:70]: Digging into the failures on `distributed/tests/test_system_monitor.py::test_disk_config`, it looks like this exception gets raised when `psutil` doesn't support the given platform/kernel (in our case, 5.6.3 didn't have support for Linux 5.5+).

Given this knowledge, would it make sense to raise a warning here to let the user know that I/O monitoring will be disabled because of this?

Comment (jrbourbeau) [reply distributed/system_monitor.py:70]: Nice detective work! I could definitely see how emitting a warning here could help avoid user confusion. Normally, I'd defer to you as to if you'd like to include that new warning/test here or in a separate PR. But given the diff here is already pretty big, and adding a new warning seems more like an (reasonable) enhancement rather than directly related to the mindeps CI build, how about we tackle it in a separate PR?

**Comment by fjetter** [comment]
I took the liberty of fixing merge conflicts in requirements.txt after merging https://github.com/dask/distributed/pull/7286

@charlesbluca is there anything else left to do?

**Comment by charlesbluca** [comment]
Thanks @fjetter!

> is there anything else left to do?

Yeah, there are two remaining tests that seem to be failing consistently with mindeps that I haven't been able to debug:

- `distributed/deploy/tests/test_spec_cluster.py::test_adaptive_killed_worker`
- `distributed/deploy/tests/test_spec_cluster.py::test_broken_worker`

Beyond these failures, it would probably be worthwhile to dig through the files that pytest is currently skipping to see if there are any tests that can be run with minimal dependencies to increase the testing coverage, but considering the diff here is pretty large I'm okay with pushing that off to a follow-up PR.

**Comment by fjetter** [comment]
> Yeah, there are two remaining tests that seem to be failing consistently with mindeps that I haven't been able to debug:

I think both test cases have the same root cause. This reminds me of failures I encountered in https://github.com/dask/distributed/pull/7323 I suggest to merge that one first (still WIP) and then see if the issue persists. I don't think this is necessarily related to the mindeps but maybe the dependencies introduced different timings making the race conditions I'm fixing over there more likely

> Beyond these failures, it would probably be worthwhile to dig through the files that pytest is currently skipping to see if there are any tests that can be run with minimal dependencies to increase the testing coverage, but considering the diff here is pretty large I'm okay with pushing that off to a follow-up PR.

Yes, that's definitely out of scope

**Review by jrbourbeau — COMMENTED** [review]
Thanks for all your work here @charlesbluca. I'm inclined to temporarily skip/xfail those failing tests in order to get this PR in. We can then un-skip/un-xfail once https://github.com/dask/distributed/pull/7323 is in. 

Here's another issue where a mindeps build would have been useful https://github.com/dask/distributed/pull/7416

**Comment by mrocklin** [comment]
It looks like this has been going on for a while.  @charlesbluca thank you for continuing to track this.  Is there anything we can do to get some variant of this in quickly so that you don't have to work to keep this from going stale?

**Comment by charlesbluca** [comment]
> Is there anything we can do to get some variant of this in quickly so that you don't have to work to keep this from going stale?

Was going to go with @jrbourbeau's suggestion and just xfail the few remaining tests that should be resolved with #7323

File: distributed/tests/test_nanny.py
Comment (charlesbluca) [inline distributed/tests/test_nanny.py:218]: Since it wasn't immediately obvious which package(s) were causing this remaining failure, I opted to xfail when `matrix.environment == 'mindeps'` in the GHA run; this means that to run mindeps tests locally, one would need to prepend the pytest command with `MINDEPS=true`.

**Comment by fjetter** [comment]
https://github.com/dask/distributed/pull/7323 might not get finished after all. I will have another final look at it but it proved to be incredibly difficult to converge and a fix might require some significant refactoring that will be out of scope for that PR

If a single/few tests are failing on this (that are not failing elsewhere) we should probably just skip them for the mindep build. #7323 should not block this

**Review by fjetter — APPROVED** [review]
Great work. Thank you @charlesbluca

**Pull request merged**
